packages/devtools_app/lib/src/geometry.dart

```diff
@@ -0,0 +1,65 @@
+import 'ui/fake_flutter/fake_flutter.dart';
+
+abstract class LineSegment {
+  LineSegment(this.start, this.end, this.name);
+
+  final Point start;
+  final Point end;
+
+  // Name is for debugging purposes.
+  final String name;
+
+  /// Whether [this] line segment intersects [rect] along the cross axis.
+  ///
+  /// For [HorizontalLineSegment]s, this will return true if [this] line segment
+  /// fits withing the y bounds of the [rect]. For [VerticalLineSegment]s, this
+  /// will return true if [this] line segment fits within the x bounds of the
+  /// [rect].
+  bool crossAxisIntersects(Rect rect);
+
+  /// Whether [this] line segment intersects [rect], bound by both its x and y
+  /// constraints.
+  bool intersects(Rect rect) {
```

Comment:
Nit: the dartGC is fast so just create a Rect that is from
start, end and call the intersect method on Rect.